prov/rxm/src/rxm_cq.c

```diff
@@ -759,6 +761,211 @@ static int rxm_handle_remote_write(struct rxm_ep *rxm_ep,
 	return 0;
 }
 
+static inline void rxm_ep_format_atomic_resp_pkt_hdr(struct rxm_conn *rxm_conn,
+				struct rxm_tx_atomic_buf *tx_buf,
+				size_t data_len, uint32_t pkt_op,
+				enum fi_datatype datatype, uint8_t atomic_op)
+{
+	rxm_ep_format_tx_buf_pkt(rxm_conn, data_len, pkt_op, 0, 0, 0,
+				 &tx_buf->pkt);
+	tx_buf->pkt.ctrl_hdr.type = ofi_ctrl_atomic_resp;
+	tx_buf->pkt.hdr.op = pkt_op;
+	tx_buf->pkt.hdr.atomic.datatype = datatype;
+	tx_buf->pkt.hdr.atomic.op = atomic_op;
+	tx_buf->pkt.hdr.atomic.ioc_count = 0;
+}
+
+static ssize_t rxm_atomic_send_resp(struct rxm_ep *rxm_ep,
+				    struct rxm_rx_buf *rx_buf,
+				    struct rxm_tx_atomic_buf *resp_buf,
+				    ssize_t result_len, uint32_t status)
+{
+	struct rxm_deferred_tx_entry *def_tx_entry;
+	struct rxm_atomic_resp_hdr *atomic_hdr;
+	ssize_t ret;
+	ssize_t resp_len = result_len + sizeof(struct rxm_atomic_resp_hdr) +
+				sizeof(struct rxm_pkt);
+
+	resp_buf->hdr.state = RXM_ATOMIC_RESP_SENT;
+	rxm_ep_format_atomic_resp_pkt_hdr(rx_buf->conn,
+					  resp_buf,
+					  resp_len,
+					  rx_buf->pkt.hdr.op,
+					  rx_buf->pkt.hdr.atomic.datatype,
+					  rx_buf->pkt.hdr.atomic.op);
+	resp_buf->pkt.ctrl_hdr.conn_id = rx_buf->conn->handle.remote_key;
+	resp_buf->pkt.ctrl_hdr.msg_id = rx_buf->pkt.ctrl_hdr.msg_id;
+	atomic_hdr = (struct rxm_atomic_resp_hdr *) resp_buf->pkt.data;
+	atomic_hdr->status = htonl(status);
+	atomic_hdr->result_len = htonl(result_len);
+
+	ret = fi_send(rx_buf->conn->msg_ep, &resp_buf->pkt, resp_len,
```

Comment:
Is this going to be problematic? Could this generate more completions than the user might expect?

Reply:
We should be fine. It can generate a completion from the core provider, but the RxM completion handler will use the protocol state associated of the completion context to know that it should be consumed and that a RxM TX CQ entry should not be written (at the target side of atomic operation).

Reply:
@jswaro - I think there is possibly another issue here. If the endpoint is configured for selective completion, then I believe it is possible that a core provider completion will not be generated here. An fi_inject when possible may be preferred and we could immediately release the resources, and the size is too big I think we need to switch to fi_sendmsg() so that I can specify FI_COMPLETION and make sure a completion is generated. Do you agree?

Reply:
I modified the code to use fi_inject when possible and avoid a completion altogether. If the inline send size is too small, then I will use fi_sendmsg() so that I can force the completion (if FI_SELECTIVE_COMPLETION was configured for the EP) and then clean up resources when the completion occurs. Either way, a completion for the internal response still will **not** be reported to the application.

Reply:
Ok.